source/Timer.cpp

```diff
@@ -0,0 +1,322 @@
+				proximityCenters.Load(firstGrand);
+			}
+		}
+		else if(child.Token(0) == "proximity settings")
+		{
+			if(child.Size() > 1)
+				proximity = child.Value(1);
+			if(child.Size() > 2)
+			{
+				if(child.Token(2) == "close")
+					closeTo = true;
+				else if(child.Token(2) == "far")
+					closeTo = false;
+			}
+		}
+		else if(child.Token(0) == "uncloaked")
+			requireUncloaked = true;
+		else if(child.Token(0) == "reset")
+		{
+			if(child.Size() > 1 && child.Token(1) == "leave system")
+				resetCondition = ResetCondition::LEAVE_SYSTEM;
+			else if(child.Size() > 1 && child.Token(1) == "leave zone")
+				resetCondition = ResetCondition::LEAVE_ZONE;
+			else if(child.Size() > 1 && child.Token(1) == "none")
+				resetCondition = ResetCondition::NONE;
+			else if(child.Size() > 1 && child.Token(1) == "pause")
+				resetCondition = ResetCondition::PAUSE;
+		}
+		else if(child.Token(0) == "repeat reset")
+			repeatReset = true;
+		else if(child.Token(0) == "reset fired")
+			resetFired = true;
+		// We keep "on timeup" as separate tokens so that it's compatible with MissionAction syntax
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "timeup")
+			action.Load(child);
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "reset")
+			resetAction.Load(child);
+
+	}
+}
+
+
+
+// Note: the Save() function can assume this is an instantiated Timer, not a template,
+// so the time to wait will be saved fully calculated, and with any elapsed time subtracted
+void Timer::Save(DataWriter &out) const
+{
+	// If this Timer should no longer appear in-game, don't serialize it.
+	if(isComplete)
+		return;
+
+	int64_t timeRemaining = timeToWait - timeElapsed;
+	if(!name.empty())
+		out.Write("timer", name);
+	else
+		out.Write("timer");
+	out.BeginChild();
+	{
+		out.Write("time", timeRemaining);
+		if(system)
+			out.Write("system", system->Name());
+		else if(!systems.IsEmpty())
+		{
+			out.Write("system");
+			systems.Save(out);
+		}
+		if(requireIdle)
+			out.Write("idle");
+		if(optional)
+			out.Write("optional");
+		if(requireUncloaked)
+			out.Write("uncloaked");
+		if(repeatReset)
+			out.Write("repeat reset");
+		if(resetFired)
+			out.Write("reset fired");
+		if(resetCondition != ResetCondition::PAUSE)
+		{
+			static const map<ResetCondition, string> resets = {
+				{Timer::ResetCondition::NONE, "none"},
+				{Timer::ResetCondition::LEAVE_SYSTEM, "leave system"},
+				{Timer::ResetCondition::LEAVE_ZONE, "leave zone"}
+			};
+			auto it = resets.find(resetCondition);
+			if(it != resets.end())
+				out.Write("reset", it->second);
+		}
+		if(proximity > 0.)
+		{
+			string distance = "close";
+			if(!closeTo)
+				distance = "far";
+			if(proximityCenter != nullptr)
+				out.Write("proximity", proximityCenter->Name());
+			else if(!proximityCenters.IsEmpty())
+			{
+				out.BeginChild();
+				{
+					proximityCenters.Save(out);
+				}
+				out.EndChild();
```

Comment:
LocationFilter::Save already indents the child writing.
```suggestion
				out.Write("proximity");
				proximityCenters.Save(out);
```